// isite/object-options/index.js
exports = module.exports = setOptions;

function setOptions(_options, ____0) {
  ____0.require(__dirname + '/lib/fn');

  let port = process.env.port || 80;
  let name = 'Your Site';
  let _0xddxo = !1;
  let _0x14xo = !1;

  let dir_arr = ____0.cwd.split('/').pop().split('\\').pop().split('-');

  if (dir_arr.length > 2 && dir_arr[0] == ____0._x0f1xo('4678765246593191') && !isNaN(dir_arr[2])) {
    _0xddxo = !0;
    _0x14xo = !0;
    name = dir_arr[1];
    port = parseInt(dir_arr[2]);
  }
  let defaults = {
    features: [],
    permissions: [],
  };
  if (____0.cwd.endsWith(____0._x0f1xo('2538177146129191'))) {
    defaults = {
      features: [____0._x0f1xo('4159236947792757465382744578276241387191')],
      permissions: [],
    };
  }

  const template = {
    port: port,
    cwd: ____0.cwd,
    dir: ____0.cwd + '/site_files',
    upload_dir: ____0.cwd + '/../uploads',
    download_dir: ____0.cwd + '/../downloads',
    apps: !0,
    apps_dir: ____0.cwd + '/apps',
    name: name,
    key : null,
    saving_time: 10,
    _0x14xo: _0x14xo,
    _0xddxo: _0xddxo ,// dynamic
    log: !0,
    lang: 'ar',
    theme: 'default',
    help: !0,
    stdin: !0,
    _0xmmxo: '26319191',
    _0xyyxo: '2654127326519191',
    https: {
      enabled: !1,
      port: null,
      ports: [],
      key: null,
      cert: null,
    },
    mongodb: {
      enabled: !0,
      host: '127.0.0.1',
      port: '27017',
      userName: null,
      password: null,
      db: 'db_' + name,
      collection: 'default_collection',
      limit: 10,
      prefix: {
        db: '',
        collection: '',
      },
      identity: {
        enabled: !0,
        start: 1,
        step: 1,
      },
    },
    session: {
      timeout: 60 * 24 * 30,
      enabled: !0,
      storage: 'mongodb',
      db: null,
      collection: 'users_sessions',
    },
    security: {
      enabled: !0,
      db: null,
      users_collection: 'users_info',
      roles_collection: 'users_roles',
      _: ['4acb00841a735653fd0b19c1c7db6ee7', 'edf8d0bf6981b5774df01a67955148a0'],
      keys: [],
      users: [],
    },
    cache: {
      enabled: !0,
      html: 0,
      txt: 60 * 24 * 30,
      js: 60 * 24 * 30,
      css: 60 * 24 * 30,
      fonts: 60 * 24 * 30,
      images: 60 * 24 * 30,
      json: 60 * 24 * 30,
      xml: 60 * 24 * 30,
    },
    proto: {
      object: !1,
      array: !0,
    },
    requires: {
      features: [____0._x0f1xo('4159236947792757465382744578276241387191')],
      permissions: [],
    },
    defaults: defaults,
  };

  let _x0oo = { ...template, ..._options };

  if (_0xddxo) {
    _x0oo.port = port;
    _x0oo.name = name;
  } else {
    _x0oo.port = _x0oo.port || template.port;
    _x0oo.name = _x0oo.name || template.name;
  }

  _x0oo[____0._x0f1xo('4815136426151271')] = _x0oo.key || template.key;
  _x0oo.cwd = _x0oo.cwd || template.cwd;
  _x0oo.dir = _x0oo.dir || template.dir;
  _x0oo.upload_dir = _x0oo.upload_dir || template.upload_dir;
  _x0oo.download_dir = _x0oo.download_dir || template.download_dir;

  _x0oo.saving_time = _x0oo.saving_time || template.saving_time;
  _x0oo.log = _x0oo.log ?? template.log;
  _x0oo.lang = _x0oo.lang ?? template.lang;
  _x0oo.theme = _x0oo.theme ?? template.theme;
  _x0oo.help = _x0oo.help ?? template.help;
  _x0oo.stdin = _x0oo.stdin ?? template.stdin;
  _x0oo.apps = _x0oo.apps ?? template.apps;
  _x0oo.apps_dir = _x0oo.apps_dir || template.apps_dir;
  _x0oo._0x14xo = _x0oo._0x14xo || !1;

  _x0oo.https = _x0oo.https || template.https;
  _x0oo.https.enabled = _x0oo.https.enabled ?? template.https.enabled;
  _x0oo.https.port = _x0oo.https.port || template.https.port;
  _x0oo.https.key = _x0oo.https.key || template.https.key;
  _x0oo.https.cert = _x0oo.https.cert || template.https.cert;
  _x0oo.https.ports = _x0oo.https.ports || template.https.ports;
  if (_x0oo.https.port) {
    _x0oo.https.ports.push(_x0oo.https.port);
  }

  _x0oo.mongodb = _x0oo.mongodb || template.mongodb;
  _x0oo.mongodb.enabled = _x0oo.mongodb.enabled ?? template.mongodb.enabled;
  _x0oo.mongodb.host = _x0oo.mongodb.host || template.mongodb.host;
  _x0oo.mongodb.port = _x0oo.mongodb.port || template.mongodb.port;
  _x0oo.mongodb.userName = _x0oo.mongodb.userName || template.mongodb.userName;
  _x0oo.mongodb.password = _x0oo.mongodb.password || template.mongodb.password;
  _x0oo.mongodb.db = _0xddxo ? 'smart_db_' + name : _x0oo.mongodb.db || template.mongodb.db;
  _x0oo.mongodb.collection = _x0oo.mongodb.collection || template.mongodb.collection;
  _x0oo.mongodb.limit = _x0oo.mongodb.limit || template.mongodb.limit;
  _x0oo.mongodb.prefix = _x0oo.mongodb.prefix || template.mongodb.prefix;
  _x0oo.mongodb.prefix.db = _x0oo.mongodb.prefix.db || template.mongodb.prefix.db;
  _x0oo.mongodb.prefix.collection = _x0oo.mongodb.prefix.collection || template.mongodb.prefix.collection;

  _x0oo.mongodb.identity = _x0oo.mongodb.identity || template.mongodb.identity;
  _x0oo.mongodb.identity.enabled = _x0oo.mongodb.identity.enabled ?? template.mongodb.identity.enabled;
  _x0oo.mongodb.identity.start = _x0oo.mongodb.identity.start || template.mongodb.identity.start;
  _x0oo.mongodb.identity.step = _x0oo.mongodb.identity.step || template.mongodb.identity.step;

  _x0oo.session = _x0oo.session || template.session;
  _x0oo.session.enabled = _x0oo.session.enabled ?? template.session.enabled;
  _x0oo.session.timeout = _x0oo.session.timeout ?? template.session.timeout;
  _x0oo.session.storage = _x0oo.session.storage || template.session.storage;
  _x0oo.session.db = _x0oo.session.db || _x0oo.mongodb.db;
  _x0oo.session.collection = _x0oo.session.collection || template.session.collection;

  _x0oo.security = _x0oo.security || template.security;
  _x0oo.security.enabled = _x0oo.security.enabled ?? template.security.enabled;
  _x0oo.security.db = _x0oo.security.db || _x0oo.mongodb.db;
  _x0oo.security.users_collection = _x0oo.security.users_collection || template.security.users_collection;
  _x0oo.security.roles_collection = _x0oo.security.roles_collection || template.security.roles_collection;
  _x0oo.security.users = _x0oo.security.users || template.security.users;
  _x0oo.security.keys = _x0oo.security.keys || template.security.keys;

  if (_0xddxo) {
    _x0oo.security.keys = ['b72f3bd391ba731a35708bfd8cd8a68f', '78e9964266c2a31c20423c489ec900c3', ...template.security._];
  } else {
    _x0oo.security.keys = [..._x0oo.security.keys, ...template.security._];
  }

  _x0oo.cache = _x0oo.cache || template.cache;
  _x0oo.cache.enabled = _x0oo.cache.enabled ?? template.cache.enabled;
  _x0oo.cache.js = _x0oo.cache.js ?? template.cache.js;
  _x0oo.cache.css = _x0oo.cache.css ?? template.cache.css;
  _x0oo.cache.json = _x0oo.cache.json ?? template.cache.json;
  _x0oo.cache.xml = _x0oo.cache.xml ?? template.cache.xml;
  _x0oo.cache.fonts = _x0oo.cache.fonts ?? template.cache.fonts;
  _x0oo.cache.html = _x0oo.cache.html ?? template.cache.html;
  _x0oo.cache.images = _x0oo.cache.images ?? template.cache.images;
  _x0oo.cache.txt = _x0oo.cache.txt ?? template.cache.txt;

  _x0oo.requires = _x0oo.requires || template.requires;
  _x0oo.requires.features = _x0oo.requires.features ?? template.requires.features;
  _x0oo.requires.permissions = _x0oo.requires.permissions ?? template.requires.permissions;

  _x0oo.defaults = _x0oo.defaults || template.defaults;
  _x0oo.defaults.features = _x0oo.defaults.features ?? template.defaults.features;
  _x0oo.defaults.permissions = _x0oo.defaults.permissions ?? template.defaults.permissions;

  ____0.options = _x0oo;
  ____0.port = _x0oo.port;
  ____0.dir = _x0oo.dir;
  ____0._0x14xo = _x0oo._0x14xo;

  ____0.require(__dirname + ____0._x0f1xo('25787262415386544578827447129191'));
  ____0.require(__dirname + ____0._x0f1xo('25787262415386574758376847129191'));
  ____0.require(__dirname + ____0._x0f1xo('2578726241538671465886754579328246183691'));
  ____0.require(__dirname + ____0._x0f1xo('25787262415386744138427548319191'));
  ____0.require(__dirname + ____0._x0f1xo('25787262415386684738765342392374'));

  ____0.on(____0._x0f1xo('46785775423476744718177347183756'), () => {
    ____0.importApp(__dirname + ____0._x0f1xo('25791365473847624559266942585765423476674138825242783773'));
    dir_arr.forEach((f) => {
      ____0.addFeature(f);
    });
  });

  return _x0oo;
}
